docs/adr/0007-dependent-deletion.md

```diff
@@ -0,0 +1,90 @@
+# 7. Dependent Deletion
+
+Date: 2019-01-20
+
+## Context
+
+Our database has a variety of inter-related tables. When we delete something from the database, we can use a couple of methods to help preserve the integrity of the remaining data: *foreign key actions* and *ActiveRecord dependencies*.
+
+Both of these methods can leave behind unused records: foreign key actions don't propagate across join tables; and a chain of ActiveRecord `dependent: :destroy` callbacks can fail halfway through (the code is not atomic).
+
+### Foreign Key Actions
+
+At the database level, we make use of foreign keys to help maintain integrity between related tables e.g.
+
+```
+# ensure creator_id on a document refers to a user in users
+# when we delete a user, unset creator_id on its documents (if any)
+add_foreign_key "documents", "users", column: "creator_id", on_delete: :nullify
+```
+
+```
+# ensure edition_id on a revision refers to an edition in editions
+# when we delete an edition, also delete all of its revisions
+add_foreign_key "revisions", "editions", column: "edition_id", on_delete: :cascade
+```
+
+```
+# ensure current_revision_id on an edition refers to a revision in revisions
+# when we delete a revision, error if its the current revision of an edition
+add_foreign_key "edition", "revision", column: "current_revision_id", on_delete: :restrict
+```
+
+Note these examples are for illustration only - they may not reflect the current schema of the app.
+
+### ActiveRecord Callbacks
+
+At the application level, we make use of the ActiveRecord `dependent` callbacks in each of the models e.g.
+
+```
+# document.rb
+# when we delete a document, run a DELETE on its timeline entries (fast)
+has_many :timeline_entries, dependent: :delete_all
+```
+
+```
+# document.rb
+# when we delee a document, invoke #destroy on each of its images (slow)
+has_many :images, dependent: :destroy
+```
+
+```
+# document.rb
+has_many :editions, dependent: :restrict_with_exception
+```
+
+Using `dependent: :destroy` is a easy way to delete a tree of records, but can also fail half-way through.
+
+## Decision
+
+We currently have no production use-case to delete anything in Content Publisher. This is sometimes necessary in development, but does not justify maintaining a bunch of `dependent` callbacks in production code.
+
+1. We will not use ActiveRecord `dependent` callbacks until we have a use-case for them.
+
+2. We will adopt a policy for foreign key actions on deletion, with the following guidelines:
+
+   * **Restrict deletion of a *more important* thing by a *less important* thing**
+      * Example: deleting an *edition* should not delete its *document*
+      * Example: deleting a *timeline entry* should not delete its *document*
```

Comment:
True. I was hoping to give abstract examples that don't require the current schema to understand them, but these also make sense and won't cause confusion if someone does look in there. Thanks :-).

Reply:
That said, I am trying to avoid words like 'invalidate', as it seems a little undefined for someone trying to follow this as a set of guidelines e.g. you could nullify the column and it would also be valid, so there's some deeper thinking going on here...